clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -0,0 +1,638 @@
+-- These are the result of the 'splitCastWork' transformation.
+inlineCast :: HasCallStack => NormRewrite
+inlineCast = inlineBinders test
+  where
+    test _ (_, (Cast (stripTicks -> Var {}) _ _)) = return True
+    test _ _ = return False
+{-# SCC inlineCast #-}
+
+-- | Given a function in the desired normal form, inline all the following
+-- let-bindings:
+--
+-- Let-bindings with an internal name that is only used once, where it binds:
+--   * a primitive that will be translated to an HDL expression (as opposed to
+--     a HDL declaration)
+--   * a projection case-expression (1 alternative)
+--   * a data constructor
+--   * I/O actions
+inlineCleanup :: HasCallStack => NormRewrite
+inlineCleanup (TransformContext is0 _) (Letrec binds body) = do
+  prims <- Lens.use (extra.primitives)
+  -- For all let-bindings, count the number of times they are referenced.
+  -- We only inline let-bindings which are referenced only once, otherwise
+  -- we would lose sharing.
+  let is1       = extendInScopeSetList is0 (map fst binds)
+      bindsFvs  = map (\(v,e) -> (v,((v,e),countFreeOccurances e))) binds
+      allOccs   = List.foldl' (unionVarEnvWith (+)) emptyVarEnv
+                $ map (snd.snd) bindsFvs
+      bodyFVs   = Lens.foldMapOf freeLocalIds unitVarSet body
+      (il,keep) = List.partition (isInteresting allOccs prims bodyFVs)
+                                 bindsFvs
+      keep'     = inlineBndrsCleanup is1 (mkVarEnv il) emptyVarEnv
+                $ map snd keep
+
+  if | null il -> return  (Letrec binds body)
+     | null keep' -> changed body
+     | otherwise -> changed (Letrec keep' body)
+  where
+    -- Determine whether a let-binding is interesting to inline
+    isInteresting
+      :: VarEnv Int
+      -> CompiledPrimMap
+      -> VarSet
+      -> (Id,((Id, Term), VarEnv Int))
+      -> Bool
+    isInteresting allOccs prims bodyFVs (id_,((_,(fst.collectArgs) -> tm),_))
+      -- Try to keep user defined names, but inline names generated by GHC or
+      -- Clash. For example, if a user were to write:
+      --
+      --    x = 2 * y
+      --
+      -- Even if 'x' is only used once, we'd like to keep it around to produce
+      -- more readable HDL. In contrast, if a user were to write:
+      --
+      --    let x = f (2 * y)
+      --
+      -- ANF would transform that to:
+      --
+      --    let x = f f_arg; f_arg = 2 * y
+      --
+      -- In that case, there's no harm in inlining f_arg.
+      | nameSort (varName id_) /= User
+      , id_ `notElemVarSet` bodyFVs
+      = case tm of
+          Prim pInfo
+            | let nm = primName pInfo
+            , Just (extractPrim -> Just p@(BlackBox {})) <- HashMap.lookup nm prims
+            , TExpr <- kind p
+            , Just occ <- lookupVarEnv id_ allOccs
+            , occ < 2
+            -> True
+            | otherwise
+            -> primName pInfo `elem` ["Clash.Explicit.SimIO.bindSimIO#"]
```

Comment:
`show 'term`

Reply:
This one I can't do: `bindSimIO#` is not imported, and there isn't a `Clash.Explicit.SimIO.Internal` module or anything